Implement support for rust-version field in project metadata

djc opened the pull request:
Needs a bunch more work, but I'd like some early feedback! Remaining work:

- [x] Bikeshedding name (picked `rust-version` here over `msrv` or `min-rust-version`)
- [x] Failing test for local dependency with unsatisfied MSRV
- [x] Requirement cannot be smaller than 1.27
- [x] Requirement cannot be smaller than initial release of edition used
- [x] Check for `run`, `verify` and `publish` subcommands
- [x] More tests (would love feedback on this)
- [x] Handle pre-release identifiers
- [x] Disallow semver range operators
- [x] Feature gate it
- [x] Add documentation for unstable feature

Minimal version of @moxian's earlier take in #7801.

cc rust-lang/rust#65262

rust-highfive commented:
r? @ehuss

(rust_highfive has picked a reviewer for you, use r? to override)

ehuss commented:
I don't have strong opinions about the name.  `rust-version` sounds reasonable to me.

As for dependency support, I think this check will need to be done elsewhere. I'm not sure if there is a perfect spot. Maybe immediately after the call to `build_unit_dependencies` it can walk the unit graph and check each package. Or it could walk the unit graph in `Context::compile` (similar to the `check_collisions` check).  I think earlier might be better?

Other things to add to the todo list is:
- Feature-gate it.
- Add documentation (`unstable.md`).

I'm a bit confused why this is changing the Summary. Shouldn't it just be stored in the Package?

The RFC mentioned "semver format without range operators". Do you happen to remember why?

Unfortunately just checking a versionreq doesn't work well with pre-release identifiers like `-nightly`. rust-version="1.40" doesn't match those (the nightly/beta tests are failing).

I can't think of a lot more tests to add. The only thing I can think of is to add some registry packages to exercise that. It looks like you added the field to the `cargo_test_support::registry::Package` test structure, so I assume you intend to follow-through with that? In `Package::publish` it will need to add the field to `Cargo.toml`. Let me know if you have any questions about that.

est31 commented:
> Requirement cannot be smaller than 1.27

Why is this the case?

djc commented:
From the RFC:

> Version can not be smaller than 1.27 (version in which package.rust field became a warning instead of an error).

djc commented:
@ehuss reverted the changes to `Summary` and `RegistryPackage` in favor of exposing this value through the `Package`, as you suggested. The current version handles the local dependency case, at least. Updated the task list.

I added the feature-gating, too, but I'm currently unsure how to then enable this feature for the tests. Any help there appreciated.

ehuss commented:
In the tests, you need to call `.masquerade_as_nightly_cargo()` on each `p.cargo(…)` invocation.  This makes cargo believe it is on the nightly channel.

For the manifest, you need to add `cargo-features = ["rust-version"]` at the top of every `Cargo.toml`.  This is the nightly-only opt-in mechanism for new fields.

bors commented:
:umbrella: The latest upstream changes (presumably #8073) made this pull request unmergeable. Please resolve the merge conflicts.

bors commented:
:umbrella: The latest upstream changes (presumably #8068) made this pull request unmergeable. Please resolve the merge conflicts.

djc commented:
@ehuss this seems like a decent MVP to merge. Please have another look.

ehuss commented on src/cargo/ops/cargo_compile.rs:
```diff
@@ -494,6 +499,29 @@ pub fn create_bcx<'a, 'cfg>(
+                    anyhow::bail!(
+                        "package `{}` cannot be built because it requires rustc {}, while \
+                         the currently active rustc version is {}",
+                        unit.pkg.name(),
```
It might be best to use `unit.pkg`, so if there are multiple instances of the same name, it can be clear which one it is referring to.

djc replied:
Fixed.

ehuss commented on src/cargo/util/command_prelude.rs:
```diff
@@ -492,6 +499,7 @@ pub trait ArgMatchesExt {
             target_rustc_args: None,
             local_rustdoc_args: None,
             rustdoc_document_private_items: false,
+            honor_rust_version: !self._is_present("ignore-rust-version"),
```
This should probably generate an error if it is not on nightly.

djc replied:
I'm not sure how to do that -- any hints?

djc replied:
Fixed.

ehuss commented on tests/testsuite/rust_version.rs:
```diff
@@ -0,0 +1,143 @@
+        .with_status(101)
+        .with_stderr(
+            "\
+error: package `bar` cannot be built because it requires rustc ^1.2345.0, while the currently active rustc version is [..]
```
Hm, the caret here might be confusing for some people, particularly those not very familiar with Cargo or semver requirements.  Perhaps if the message could say something like "... it requires rustc 1.2345.0 or newer, ...".

djc replied:
Fixed. This required a bit of surgery, with the result that the manifest propagates the value as a string, which gets checked for correctness (including parsing as `VersionReq`) when the manifest gets ingested, but is parsed again when checking in cargo_compile, where I `unwrap()` the error (since this is checked earlier). Is that okay?

ehuss replied:
Yea, I think so.  If it's a problem later, it can be changed.

ehuss commented on tests/testsuite/rust_version.rs:
```diff
@@ -0,0 +1,143 @@
+//! Tests for targets with `rust-version`.
```
I'd like to see the following tests:

- That it is feature gated (see all the *_gated tests).
- How it works with registry dependencies.
- Bad syntax for `rust-version`.

djc replied:
Added a `_gated` test and tests for bad syntax. Any suggestion for a registry dependency test I can crib off of?

ehuss replied:
Looking closer, I would just change `rust_version_dependency_fails` to use `Package::new` to create the `bar` dependency instead of using a path dependency. This is just to check the behavior when it is used as a registry dependency. Path and registry dependencies aren't all that different, and I don't expect there to be any problems, just seems nice to verify.

I can't think of any tests that are exactly the same, but [this](https://github.com/rust-lang/cargo/blob/a73e5b7d567c3036b296fc6b33ed52c5edcd882e/tests/testsuite/features_namespaced.rs#L99-L137) is similar. There is some documentation and examples on how to use `Package::new` [here](https://github.com/rust-lang/cargo/blob/a73e5b7d567c3036b296fc6b33ed52c5edcd882e/crates/cargo-test-support/src/registry.rs#L73-L134).

djc replied:
Done.

ehuss commented on src/cargo/util/toml/mod.rs:
```diff
@@ -1028,6 +1030,25 @@ impl TomlManifest {
+        let rust_version = project
+            .rust_version
+            .as_ref()
+            .map(|v| semver::Version::parse(&v))
```
Would it be possible to parse partial versions?  Like `1.43`.  I think 99% of the time nobody is going to care about the patch version.  (I think it was also stipulated in the rfc.)

djc replied:
This is supported now, and is tested in one of the tests.

bors commented:
:umbrella: The latest upstream changes (presumably #8167) made this pull request unmergeable. Please resolve the merge conflicts.

oxalica commented:
Don't we need check for `update` sub-command?
I think we should avoid bumping dependencies to versions that break current `rust-verion`.

djc commented:
That was explicitly scoped out of RFC 2495.

ehuss commented:
Ping @djc:  do you think you'll be able to address the comments and rebase on master?

djc commented on src/doc/src/reference/unstable.md:
```diff
@@ -747,3 +747,13 @@ The following is a description of the JSON structure:
+* RFC: [#2495](https://github.com/rust-lang/rfcs/blob/master/text/2495-min-rust-version.md)
+* rustc Tracking Issue: [#65262](https://github.com/rust-lang/rust/issues/65262)
+
+The `-Z rust-version` flag enables the reading the `rust-version` field in the
```
Can you tell me how to wire it up correctly?

djc replied:
I think this is done now.

djc commented on src/cargo/core/features.rs:
```diff
@@ -430,6 +444,7 @@ impl CliUnstable {
             "features" => self.features = Some(parse_features(v)),
             "crate-versions" => self.crate_versions = parse_empty(k, v)?,
             "separate-nightlies" => self.separate_nightlies = parse_empty(k, v)?,
+            "rust-version" => self.rust_version = parse_bool(k, v)?,
```
Fixed.

djc commented on src/cargo/ops/cargo_compile.rs:
```diff
@@ -494,6 +499,29 @@ pub fn create_bcx<'a, 'cfg>(
 
+    if honor_rust_version {
+        let current_version = &target_data.rustc.version;
+        let untagged_version = semver::Version::new(
```
Done.

djc commented on src/doc/src/reference/unstable.md:
```diff
@@ -747,3 +747,13 @@ The following is a description of the JSON structure:
+### rust-version
+* RFC: [#2495](https://github.com/rust-lang/rfcs/blob/master/text/2495-min-rust-version.md)
+* rustc Tracking Issue: [#65262](https://github.com/rust-lang/rust/issues/65262)
+
```
Have another look?

djc commented on src/cargo/util/command_prelude.rs:
```diff
@@ -207,6 +207,13 @@ pub trait AppExt: Sized {
+        self._arg(opt(
+            "ignore-rust-version",
+            "Ignore `rust-version` specification in packages",
+        ))
```
Done.

djc commented on src/cargo/util/command_prelude.rs:
```diff
@@ -207,6 +207,13 @@ pub trait AppExt: Sized {
         self._arg(opt("dry-run", dry_run))
     }
+
+    fn arg_ignore_rust_version(self) -> Self {
```
I'm not so sure how this is supposed to be wired up. Can you point me in the right direction?

djc replied:
Added the missing commands.

djc commented on src/cargo/util/toml/mod.rs:
```diff
@@ -771,6 +771,8 @@ impl<'de> de::Deserialize<'de> for VecStringOrBool {
 #[derive(Deserialize, Serialize, Clone, Debug)]
 pub struct TomlProject {
     edition: Option<String>,
+    #[serde(rename = "rust-version")]
```
Done.

djc commented on src/cargo/core/manifest.rs:
```diff
@@ -540,6 +543,17 @@ impl Manifest {
+                .chain_err(|| {
+                    anyhow::format_err!(
+                        "the 'rust-version' manifest key is unstable and may \
+                         erroneously fail to compile your crates"
```
Removed the extra error message.

djc commented on src/cargo/util/toml/mod.rs:
```diff
@@ -1028,6 +1030,25 @@ impl TomlManifest {
+            .rust_version
+            .as_ref()
+            .map(|v| semver::Version::parse(&v))
+            .transpose()?;
```
Added.

djc commented on src/cargo/core/features.rs:
```diff
@@ -352,6 +365,7 @@ pub struct CliUnstable {
     pub features: Option<Vec<String>>,
     pub crate_versions: bool,
     pub separate_nightlies: bool,
+    pub rust_version: bool,
```
Should it be? If so, how/where?

djc replied:
Removed.

djc commented on crates/cargo-test-support/src/registry.rs:
```diff
@@ -361,6 +363,12 @@ impl Package {
 
+    /// Specify a minimal Rust version.
+    pub fn rust_version(&mut self, rust_version: &str) -> &mut Package {
+        self.rust_version = Some(rust_version.into());
```
If you point me at some example tests, I can add some new ones for this.

djc replied:
Fixed.

ehuss commented:
I think for this, we don't need a specific `-Z` flag, so I think `rust_version` can be removed from `CliUnstable`. Just when the user passes the `--ignore-rust-version` flag, it should call `fail_if_stable_opt` in `command_prelude.rs` to ensure the user also passed in `-Z unstable-options`. Search around for some examples of using that method.

For nightly/unstable features, there are two independent opt-in mechanisms. The `cargo-features = […]` in `Cargo.toml` is for allowing unstable syntax in `Cargo.toml`. This is checked by the `features.required(Feature::rust_version())` which you have correctly added. The command-line `-Z` flags are only needed for CLI extensions (things that don't appear in `Cargo.toml`). How those are used depend on the situation:

* If there is just a new flag, like `--config`, then you call [`config.unstable_flags.fail_if_stable_opt(...)`](https://github.com/rust-lang/cargo/blob/a73e5b7d567c3036b296fc6b33ed52c5edcd882e/src/cargo/util/config/mod.rs#L714) to ensure the user also passed in `-Z unstable-options` to enable it.
* If there is new behavior that otherwise doesn't have any new CLI options, then we add a new `-Z` flag to the `CliUnstable` struct, and then check [`config.cli_unstable().my_feature_name`](https://github.com/rust-lang/cargo/blob/a73e5b7d567c3036b296fc6b33ed52c5edcd882e/src/cargo/ops/cargo_compile.rs#L364). Unstable `-Z` flags are already rejected on non-nightly channels.

There is a bit of documentation on this at https://doc.crates.io/contrib/process/unstable.html and in the [features module](https://github.com/rust-lang/cargo/blob/master/src/cargo/core/features.rs).

----

As for centralizing the argument, there are a set of arguments that are essentially the same for every "build-like" command. That is:

* bench
* build
* check
* doc
* fix
* run
* rustc
* rustdoc
* test

So we don't have to fuss with ensuring that these commands always include the right options, I was thinking it would be nice just to have an `arg_build_common()` method in `command_prelude.rs` that adds those options. It should call:

* `arg_features`
* `arg_jobs`
* `arg_target_triple`
* `arg_target_dir`
* `arg_manifest_path`
* `arg_message_format`
* `arg_ignore_rust_version`

It's up to you if you want to do that. Otherwise, just make sure the list above is covered by `arg_ignore_rust_version`. I think adding a common method would help when we add more flags in the future that need to be included in all of these commands.

djc commented:
I have not grouped the build common options in this PR, I might do a follow-up to improve the situation there.

djc commented:
From the RFC:

> Version can not be smaller than 1.27 (version in which package.rust field became a warning instead of an error).

I don't know if/how to implement this. I think Cargo will now still error if the unstable field gets used. I could restrict it for now to not allow it to be smaller than 1.51, if that makes sense, but that version should eventually probably be adjusted to the version that stabilizes this field.

ehuss commented on crates/cargo-test-support/src/registry.rs:
```diff
@@ -502,15 +517,26 @@ impl Package {
+        let mut manifest = String::new();
+
+        if !self.cargo_features.is_empty() {
+            manifest.push_str(&format!("cargo-features = {:?}\n\n", self.cargo_features));
```
```suggestion
            manifest.push_str(&format!(
                "cargo-features = {}\n\n",
                toml::to_string(&self.cargo_features).unwrap()
            ));
```

Although it is unlikely to ever change, I feel uncomfortable using debug formatting like this.  (This will require adding `toml = "0.5"` to `cargo-test-support/Cargo.toml`.)

ehuss commented on crates/cargo-test-support/src/registry.rs:
```diff
@@ -502,15 +517,26 @@ impl Package {
+        ));
+
+        if let Some(version) = &self.rust_version {
+            manifest.push_str(&format!("rust-version = {:?}", version));
```
```suggestion
            manifest.push_str(&format!("rust-version = \"{}\"", version));
```

ehuss commented on src/cargo/core/manifest.rs:
```diff
@@ -504,6 +507,10 @@ impl Manifest {
                 })?;
         }
 
+        if self.rust_version.is_some() {
+            self.unstable_features.require(Feature::rust_version())?;
+        }
+
```
```suggestion
```

I don't think this is necessary, since it is already checked in `to_real_manifest`.

ehuss commented:
Thanks! 

As for the 1.27 floor, I think the intent is: let's say it gets stabilized in version 1.60.  A project whose MSRV is 1.35 can specify 1.35 in the manifest to make it clear what their MSRV is.  Even though older versions will ignore it, I think it makes sense to specify the actual version instead of the version when it first starts getting honored.

However, thinking about that more, I think if this gets merged now, that means that versions starting with 1.51 will never be able to use this setting, because Cargo will complain about an unstable feature.  I'm concerned that this will make it very difficult to use for a long time (until a project's MSRV reaches whenever it is stabilized, which could be a year or more).

Since this feature is fundamentally about improving MSRV support, I think that may be a bad way to start it off.  This key is not critical (it doesn't actually change anything).  I'm thinking instead of calling `unstable_features.require(...)` in toml/mod.rs it could instead do something like this:

```rust
if features.require(Feature::rust_version()).is_err() {
    warnings.push(
        "`rustc-version` is not supposed on this version of Cargo and will be ignored"
            .to_string(),
    );
}
```

Maybe include a check to `nightly_features_allowed()` and add a hint about how to enable it on nightly (similar to how `require()` works).

I *think* that should be safe to do, it will just print a warning just like any other unknown key.

djc commented:
Resolved your final comments and changed the error to warn (including tests for both nightly and stable).

ehuss commented:
Looks good, thanks! 🎉 

@bors r+

bors commented:
:pushpin: Commit c221fec91119cd94a3043cc19566ed61d4f5c13d has been approved by `ehuss`

<!-- @bors r=ehuss c221fec91119cd94a3043cc19566ed61d4f5c13d -->
<!-- homu: {"type":"Approved","sha":"c221fec91119cd94a3043cc19566ed61d4f5c13d","approver":"ehuss"} -->

bors commented:
:hourglass: Testing commit c221fec91119cd94a3043cc19566ed61d4f5c13d with merge 8e075c9cab41eb1ed6222f819924999476477f2e...
<!-- homu: {"type":"BuildStarted","head_sha":"c221fec91119cd94a3043cc19566ed61d4f5c13d","merge_sha":"8e075c9cab41eb1ed6222f819924999476477f2e"} -->

djc commented:
Thanks for the reviews!

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/cargo/runs/1736289066)
Approved by: ehuss
Pushing 8e075c9cab41eb1ed6222f819924999476477f2e to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"ehuss","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/cargo/runs/1736289066"},"merge_sha":"8e075c9cab41eb1ed6222f819924999476477f2e"} -->
